oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/ReactiveJwtProcessors.java

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.jwt;
+
+import java.security.interfaces.RSAPublicKey;
+import java.util.function.Function;
+
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JWSAlgorithm;
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jose.jwk.JWKMatcher;
+import com.nimbusds.jose.jwk.JWKSelector;
+import com.nimbusds.jose.jwk.JWKSet;
+import com.nimbusds.jose.jwk.RSAKey;
+import com.nimbusds.jose.jwk.source.ImmutableJWKSet;
+import com.nimbusds.jose.jwk.source.JWKSecurityContextJWKSet;
+import com.nimbusds.jose.jwk.source.JWKSource;
+import com.nimbusds.jose.proc.BadJOSEException;
+import com.nimbusds.jose.proc.JWKSecurityContext;
+import com.nimbusds.jose.proc.JWSKeySelector;
+import com.nimbusds.jose.proc.JWSVerificationKeySelector;
+import com.nimbusds.jose.proc.SecurityContext;
+import com.nimbusds.jwt.JWT;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import com.nimbusds.jwt.proc.DefaultJWTProcessor;
+import com.nimbusds.jwt.proc.JWTProcessor;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.security.oauth2.jose.jws.JwsAlgorithms;
+import org.springframework.util.Assert;
+import org.springframework.web.client.RestOperations;
+import org.springframework.web.reactive.function.client.WebClient;
+
+/**
+ * A collection of builders for creating {@link ReactiveJWTProcessor} instances.
+ *
+ * @author Josh Cummings
+ * @since 5.2
+ * @see NimbusJwtDecoder
+ */
+public class ReactiveJwtProcessors {
```

Comment:
After our changes I believe this name should change since there is no `ReactiveJwtProcessor` now.

I'm still thinking it might be valuable to move this into NimbusReactiveJwtDecoder. If we foresee it being used somewhere else, then we would want to keep it externalized though.

> the output of the builder would be the decoder itself as opposed to a contract similar to Nimbus's `JWTProcessor`, making this kind of reuse trickier.

I don't think this is a problem. User's would still be able to inject a `Converter` directly.

Reply:
If the builder returns a decoder and later on that builder moves out of `NimbusReactiveJwtDecoder` into its own class, the existing decoder constructors will need to be deprecated. I'm not so keen on their deprecation being tied to that refactor. 

If we can, I'd prefer that the builder return the equivalent of a processor. (And if it comes down to naming, I have some thoughts below.)

>  If we foresee it being used somewhere else ...

I think reuse is a very good point, and I struggle a bit with it because of `JwtProcessors`, which returns a `JWTProcessor`, a Nimbus-specific type. This builder has provided a lot of flexibility for `NimbusJwtDecoder`, making it easy to add other processing types without modifying the decoder. Will this pattern of building `JWTProcessor`s be applied to other places than the `NimbusJwtDecoder` constructor? I would guess the answer to this is the answer here as well.

> After our changes I believe this name should change since there is no ReactiveJwtProcessor now.

Again, this is sensible - `JwtProcessors`  builds a `JWTProcessor`, `JwtDecoders` builds a `JwtDecoder`, etc.

We have an example of a converter whose parameter types don't alone describe what the converter is for: `MappedJwtClaimSetConverter implements Converter<Map<String, Object>, Map<String, Object>>`. Here, the name clarifies the semantics of the conversion and we don't want to call it `MapMapConverter` or similar.

Also, we have classes (not interfaces, AFAIK) that return Nimbus types, for example, `JwtProcessors` returns a `JWTProcessor`.

Given these, I believe it's sensible to either give a distinctive name to a certain type of converter, in order to clarify semantics -- have the interface `ReactiveJwtProcessor extends Converter<SignedJWT, JWTClaimSet>` (read: it's like Nimbus's JWTProcessor, but reactive) -- or to simply state what we are producing, even if the builder's name for a `Converter<S,T>` doesn't precisely match `STConverter` -- do `ReactiveJwtProcessors` since what is effectively produced is a reactive version of Nimbus's JWTProcessor.

Though, given that I'm saying "Nimbus's JWTProcessor" so much, it might be better to call it `ReactiveNimbusJwtProcessors`.

This has come about because of the absence of a reactive API for Nimbus, and I'm guessing it won't be the last time. Possibly, we could gauge Nimbus's interest in maintaining a reactive version of their API that depends on Project Reactor, though that seems unlikely. Given that, I don't think it's unreasonable to introduce types that are the reactive equivalent of Nimbus APIs and calling them such.

Reply:
Had a chat offline with @rwinch and will summarize here.

The main reason for my concern is symmetry between the servlet and reactive APIs. So, since neither is released for this feature, yet, we decided to actually change the servlet version so that it also internalized the decoder builder.